lib/utils/platform/iot_threads_afr.c

```diff
@@ -0,0 +1,350 @@
+{
+    return prIotMutexTimedLock( pMutex, 0 );
+}
+
+/*-----------------------------------------------------------*/
+
+void IotMutex_Unlock( IotMutex_t * const pMutex )
+{
+    _iot_mutex_internal_t * internalMutex = ( _iot_mutex_internal_t * ) pMutex;
+
+    IotLogDebug( "Unlocking mutex %p.", internalMutex );
+    
+    /* Call the correct FreeRTOS mutex unlock function based on mutex type. */
+    if( internalMutex->recursive == true )
+    {
+        ( void ) xSemaphoreGiveRecursive( ( SemaphoreHandle_t ) & internalMutex->xMutex );
+    }
+    else
+    {
+         ( void ) xSemaphoreGive( ( SemaphoreHandle_t ) & internalMutex->xMutex );
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+bool IotSemaphore_Create( IotSemaphore_t * const pNewSemaphore,
+                             uint32_t initialValue,
+                             uint32_t maxValue )
+{
+    _iot_sem_internal_t * internalSemaphore = ( _iot_sem_internal_t * ) pNewSemaphore;
+    IotLogDebug( "Creating new semaphore %p.", pNewSemaphore );
+
+    ( void ) xSemaphoreCreateCountingStatic( maxValue, initialValue, &internalSemaphore->xSemaphore );
+
+    return true;
+}
+
+/*-----------------------------------------------------------*/
+
+uint32_t IotSemaphore_GetCount( IotSemaphore_t * const pSemaphore )
+{
+    _iot_sem_internal_t * internalSemaphore = ( _iot_sem_internal_t * ) pSemaphore;
+    UBaseType_t count = 0;
+
+    count = ( UBaseType_t ) uxSemaphoreGetCount( ( SemaphoreHandle_t ) &internalSemaphore->xSemaphore );
+
+    IotLogDebug( "Semaphore %p has count %d.", pSemaphore, count );
+
+    return ( uint32_t ) count;
+}
+
+/*-----------------------------------------------------------*/
+
+void IotSemaphore_Destroy( IotSemaphore_t * const pSemaphore )
+{
+    _iot_sem_internal_t * internalSemaphore = ( _iot_sem_internal_t * ) pSemaphore;
+
+    IotLogDebug( "Destroying semaphore %p.", internalSemaphore );
+
+    vSemaphoreDelete( ( SemaphoreHandle_t ) &internalSemaphore->xSemaphore );
+}
+
+/*-----------------------------------------------------------*/
+
+void IotSemaphore_Wait( IotSemaphore_t * pSemaphore )
+{
+    _iot_sem_internal_t * internalSemaphore = ( _iot_sem_internal_t * ) pSemaphore;
+
+    IotLogDebug( "Waiting on semaphore %p.", internalSemaphore );
+
+    /* Take the semaphore using the FreeRTOS API. */
+    if( xSemaphoreTake( ( SemaphoreHandle_t ) &internalSemaphore->xSemaphore,
+                          portMAX_DELAY ) != pdTRUE )
+    {
+        IotLogWarn( "Failed to wait on semaphore %p.",
+                       pSemaphore);
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+bool IotSemaphore_TryWait( IotSemaphore_t * const pSemaphore )
+{
+    _iot_sem_internal_t * internalSemaphore = ( _iot_sem_internal_t * ) pSemaphore;
+
+    IotLogDebug( "Attempting to wait on semaphore %p.", internalSemaphore );
+
+    return IotSemaphore_TimedWait( pSemaphore, 0 );
```

Comment:
tail cal... generally speaking for debuggability purposes, kernel and related libraries favor a pattern such as the following:


     error = IotSemaphore_TimedWait( pSemaphore, 0 );

if(error is an error)
{
   ...log...
   ...assert... (in this case we should assert (see comment above) because this should NOT fail)
}

return error;

Reply:
Agreed on the concept yes, but this function is just really implementing function overloading, the real implementation does log when a timeout happened, and if there is no timeout it always returns false, so above does not work here.